### PR TITLE
Setup jemalloc in default Dockerfile to optimize memory allocation

### PR DESCRIPTION
**Pull request by northeastprince**
[Ruby's use of `malloc` can create memory fragmentation problems, especially when using multiple threads](https://www.speedshop.co/2017/12/04/malloc-doubles-ruby-memory.html) like Puma does. Switching to an allocator that uses different patterns to avoid fragmentation can decrease memory usage by a substantial margin.

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt [hunk 1]
```diff
@@ -85,6 +85,8 @@ RUN apt-get update -qq && \
     apt-get install --no-install-recommends -y <%= dockerfile_deploy_packages.join(" ") %> && \
     rm -rf /var/lib/apt/lists /var/cache/apt/archives
 
+ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2
```
Comment (byroot) [inline railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:88]: This won't work if you build on ARM64 (typically M1 mac) or deploy to ARM servers (e.g. Graviton).

**Review by byroot — COMMENTED** [review]
I agree we should do this, as it's really a no brainer in term of performance.

However as is this PR break a few stuff and need some more work. IIRC the `jemalloc` package comes with a command that does the proper LD_PRELOAD, we should just use that.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt (hunk 1 above)
Comment (northeastprince) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt]: @byroot there doesn't seem to be a way to interpolate commands in the ENV directive - we might have to move this to the `docker-entrypoint`.

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/Dockerfile.tt]: Yeah, that's where I meant to put the `jemalloc` command. But I misremembered anyway, the package doesn't include it.

I'll add the entrypoint, it shouldn't be too hard.

**Comment by gjtorikian** [comment]
Apologies for the drive-by comment, but I think the "better way" would be to patch Ruby directly to use libjemalloc, as mastodon shows: https://github.com/mastodon/mastodon/blob/1666b1955992e16f4605b414c6563ca25b3a3f18/Dockerfile#L108-L113

`LD_PRELOAD` affects _every_ binary, not just Ruby, which may not be desirable. More details [here](https://github.com/docker-library/ruby/issues/182#issuecomment-1368432448).

**Comment by gjtorikian** [comment]
Gotchu!

File: railties/lib/rails/generators/rails/app/templates/docker-entrypoint.tt [hunk 2]
```diff
@@ -1,5 +1,10 @@
 #!/bin/bash -e
 
+# Enable jemalloc for reduced memory usage and reduce latency.
```
Comment (joshuay03) [inline railties/lib/rails/generators/rails/app/templates/docker-entrypoint.tt:3]: Minor typo:

```suggestion
# Enable jemalloc for reduced memory usage and reduced latency.
```

Or just removing is more succint imo:

```suggestion
# Enable jemalloc for reduced memory usage and latency.
```

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/docker-entrypoint.tt:3]: https://github.com/rails/rails/commit/1ca8c8ea3401ec1a88ea5a5adf56b9c65a6746a6

**Comment by suwyn** [comment]
@byroot @northeastprince I'm not seeing jemalloc being used in the process map.  Can you double check?

Compare the output of

`pmap -X 1 | grep jemalloc`

on the rails server process vs the output using 

`exec env LD_PRELOAD="$LD_PRELOAD" "${@}"`

in the entrypoint script

**Comment by byroot** [comment]
Yeah, I forgot to add `export`. Fixed in 9940dc879d9bf5188f9237341329534437ee208b. Thanks for noticing.

**Pull request merged**